Update Modules and Functors Tutorials (#1778)
ocaml/ocaml.org: data/tutorials/language/1ms_00_modules.md

# dalet_of (Some (Either.Left 10));;
- : dalet = Dennis 10

# let dalet_to_string = function
  | Dorothy -> "Dorothy"
  | Dennis _ -> "Dennis"
  | Donald _ -> "Donald";;
val dalet_to_string : dalet -> string = <fun>
```

The type `dalet` is _read-only_. Pattern matching is possible, but values can only be constructed by the provided functions, here `dalet_of`.

Abstract and read-only types can be either variants, as shown in this section, records, or aliases. It is possible to access a read-only record field's value, but creating such a record requires using a provided function.

## Submodules

### Submodule Implementation

A module can be defined inside another module. That makes it a _submodule_.
Let's consider the files `florence.ml` and `glasgow.ml`

**`florence.ml`**
```ocaml
module Hello = struct
  let message = "Hello from Florence"
  let print () = print_endline message
end

let print_goodbye () = print_endline "Goodbye"
```

**`glasgow.ml`**
```ocaml
let () =
  Florence.Hello.print ();
  Florence.print_goodbye ()
```

Definitions from a submodule are accessed by chaining module names, here
`Florence.Hello.print`.

### Submodule With Signatures

> To define an interface to a submodule, we can provide a _module signature_. This
> is done in this second version of the `florence.ml` file:

Reviewer: ```suggestion
To define a submodule's interface, we can provide a _module signature_. This
is done in this second version of the `florence.ml` file:
```

Would this still be correct?

Author: Yes better

```ocaml
module Hello : sig
 val print : unit -> unit
end = struct
  let message = "Hello"
  let print () = print_endline message
end

let print_goodbye () = print_endline "Goodbye"
```

The first version made `Florence.Hello.message` public. In this version it can't be accessed from `glasgow.ml`.

### Module Signatures are Types

> The role played by module signatures to implementations is akin to the role played by types to values. Here is a third possible way to write file `florence.ml`:

Reviewer: ```suggestion
The module signatures' role to implementations is akin to the role played by types to values. Here is a third possible way to write file `florence.ml`:
```
Is this still correct?

Author: Yes it is correct. Can we use possessive for the second part of the sentence, after “akin”, the type & value?

but we really want it as if it were part of it. In an `extlib.ml` file, we
can achieve this effect by using the `include` directive:

```ocaml
module List = struct
  include Stdlib.List
  let uncons = function
    | [] -> None
    | hd :: tl -> Some (hd, tl)
end
```

It creates a module `Extlib.List` that has everything the standard `List` module
has, plus a new `uncons` function. In order to override the default `List`
module from another `.ml` file, we need to add `open Extlib` at the beginning.

## Stateful Modules

A module may have an internal state. This is the case for the `Random` module from the standard library. The functions `Random.get_state` and `Random.set_state` provide read and write access to the internal state, which is nameless and has an abstract type.
```ocaml
# let s = Random.get_state ();;
val s : Random.State.t = <abstr>

# Random.bits ();;
- : int = 89809344

# Random.bits ();;
- : int = 994326685

# Random.set_state s;;
- : unit = ()

# Random.bits ();;
- : int = 89809344
```

Values returned by `Random.bits` will differ when you run this code. The first
and third calls return the same results, showing that the internal state was
reset.

## Conclusion

> In OCaml, modules are the basic means of organising software. To sum up, a module is a collection of definitions wrapped under a name. These definitions can be submodules, which allows the creation of hierarchies of modules. Top-level modules must be files and are the units of compilation. Every module has an interface, which is the list of definitions a module exposes. By default, a module's interface exposes all its definitions, but this can be restricted using the interface syntax.

Reviewer: ```suggestion
In OCaml, modules are the basic means of organising software. To sum up, a module is a collection of definitions wrapped under a name. These definitions can be submodules, which allow the creation of module hierarchies. Top-level modules must be files and are the units of compilation. Every module has an interface, the list of definitions a module exposes. By default, a module's interface exposes all its definitions, but this can be restricted using the interface syntax.
```

Author: Yes, that's better

itself can be a submodule of another module, pretty much like directories in a
file system.

When you write a program, let's say using the two files `athens.ml` and

> `berlin.ml`, each automatically defines a module named

Reviewer: Do they define modules or create modules (or do those both mean the same thing)?

Author: Defines is probably better.

# let array_sum_sq m =
    Array.(init m Fun.id |> map (fun i -> i * i) |> fold_left ( + ) 0);;
val array_sum_sq : int -> int = <fun>
```

## Interfaces and Implementations

> By default, anything defined in a module is accessible from other modules.
> Values, functions, types, or submodules, everything is public. This can be
> restricted. That allows distinguishing content provided to other modules from
> internal use content. What is internal is kept private and not available from
> other modules.

Reviewer: ```suggestion
By default, anything defined in a module is accessible from other modules, including
values, functions, types, or submodules. Everything is public. However, this can be
restricted, so distinguished content provided to other modules can be distinguished from
internal use content. Internal information is kept private and not available from
other modules.
```

Not sure about that second-to-last sentence. If I got it wrong, let's work on it togther.

- : string = "Christine"
```

> Type `gimel` is _abstract_. Values are available, but only as function result or argument. Only the provided functions `gimel_of_bool`, `gimel_flip`, and ` gimel_to_string` and polymorphic functions can receive or return `gimel` values.

Reviewer: Since "Values" is plural here, it is correct for it to be "function results or arguments," both plural.
